src/geom/detail/frameSetUtils.cc

```diff
@@ -0,0 +1,246 @@
+/*
+ * LSST Data Management System
+ * Copyright 2017 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include <algorithm>
+#include <exception>
+#include <memory>
+#include <ostream>
+#include <set>
+#include <sstream>
+#include <vector>
+
+#include "astshim.h"
+
+#include "lsst/afw/formatters/Utils.h"
+#include "lsst/afw/geom/Angle.h"
+#include "lsst/afw/geom/Point.h"
+#include "lsst/afw/geom/detail/frameSetUtils.h"
+#include "lsst/daf/base/PropertyList.h"
+#include "lsst/pex/exceptions.h"
+
+namespace lsst {
+namespace afw {
+namespace geom {
+namespace detail {
+namespace {
+
+// destructively make a set of strings from a vector of strings
+std::set<std::string> setFromVector(std::vector<std::string>&& vec) {
+    return std::set<std::string>(std::make_move_iterator(vec.begin()), std::make_move_iterator(vec.end()));
+}
+
+} // namespace
+
+std::shared_ptr<ast::SkyFrame> getSkyFrame(ast::FrameSet const& frameSet, int index, bool copy) {
+    auto frame = frameSet.getFrame(index, false);
+    auto skyFrame = std::dynamic_pointer_cast<ast::SkyFrame>(frame);
+    if (!skyFrame) {
+        std::ostringstream os;
+        os << "Bug! Frame at index=" << index << " is a " << frame->getClass() << ", not a SkyFrame";
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, os.str());
+    }
+    return skyFrame;
+}
+
+std::shared_ptr<daf::base::PropertyList> makeTanWcsMetadata(Point2D const& crpix, SpherePoint const& crval,
+                                                            Eigen::Matrix2d const& cdMatrix) {
+    auto pl = std::make_shared<daf::base::PropertyList>();
+    pl->add("RADESYS", "ICRS");
+    pl->add("EQUINOX", 2000);  // not needed, but may help some older code
+    pl->add("CTYPE1", "RA---TAN");
+    pl->add("CTYPE2", "DEC--TAN");
+    pl->add("CRVAL1", crval[0].asDegrees());
+    pl->add("CRVAL2", crval[1].asDegrees());
+    pl->add("CRPIX1", crpix[0] + 1);
+    pl->add("CRPIX2", crpix[1] + 1);
+    pl->add("CD1_1", cdMatrix(0, 0));
+    pl->add("CD1_2", cdMatrix(0, 1));
+    pl->add("CD2_1", cdMatrix(1, 0));
+    pl->add("CD2_2", cdMatrix(1, 1));
+    pl->add("CUNIT1", "deg");
+    pl->add("CUNIT2", "deg");
```

Comment:
(Answered earlier. This new code is intended to eventually replace the old code.)